sensors/Sensors.cpp

```diff
@@ -0,0 +1,104 @@
+#include "Sensors.h"
+#include "math/EigenWrapper.h"
+#include "utils_and_transforms.h"
+#include "ParameterParser.h"
+#include "SRP.h"
+#include "MagneticField.h"
+#include <unsupported/Eigen/MatrixFunctions>
+#include <cmath>
+
+#ifdef USE_PYBIND_TO_COMPILE
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"   // purposely comparing floats
+#include "pybind11/eigen.h"
+#pragma GCC diagnostic pop
+#endif
+
+VectorXd ReadSensors(const VectorXd state, double t_J2000, Simulation_Parameters sc)
+{
+    int measurement_vec_size = 6 + 3 + 3 + sc.num_light_diodes; // GPS + Gyro + Mag Field + Light Sensors
+    VectorXd measurement = VectorXd::Zero(measurement_vec_size);
+
+    measurement(Eigen::seqN(0,6)) = GPS(state(Eigen::seqN(0,3)), state(Eigen::seqN(3,3)), t_J2000, sc);
+    measurement(Eigen::seqN(6,3)) = Gyroscope(state(Eigen::seqN(10,3)), sc);
+    measurement(Eigen::seqN(9,3)) = Magnetometer(state(Eigen::seqN(0,3)), state(Eigen::seqN(6,4)), t_J2000, sc);
+    measurement(Eigen::seqN(12, sc.num_light_diodes)) = SunSensor(state(Eigen::seqN(6,4)), t_J2000, sc);
+
+    return measurement;
+}
+
+Vector6 GPS(const Vector3 pos, const Vector3 vel, double t_J2000, Simulation_Parameters sc)
+{
+    Vector6 y = Vector6::Zero();
+
+    Matrix_3x3 R_ECI2ECEF = ECI2ECEF(t_J2000);
+
+    Vector3 pos_noise = Vector3::Random();
+    Vector3 vel_noise = Vector3::Random();
```

Comment:
I've been burned by this before. It seems Eigen's random functions give the same value every time you re-run your program.

Reply:
Thanks for this. Eigen does return the same samples by trial. Switched to a std::normal_distribution instead